gadjid/src/lib.rs

```diff
@@ -9,3 +9,194 @@ mod partially_directed_acyclic_graph;
+
+    use rand::{Rng, SeedableRng};
+    use rustc_hash::FxHashSet;
+
+    use crate::{
+        graph_operations::{self, parent_aid},
+        PDAG,
+    };
+
+    fn load_pdag_from_mtx(full_path: &str) -> PDAG {
+        // read the mtx file
+        let mtx = std::fs::read_to_string(full_path).unwrap();
+
+        let mut lines = mtx.lines();
+
+        // skipping first line of mtx format that give metadata like dimensions
+        lines.next();
+
+        let dims = lines
+            .next()
+            .unwrap()
+            .split_whitespace()
+            .collect::<Vec<&str>>();
+        let rows = dims[0].parse::<usize>().unwrap();
+        let cols = dims[1].parse::<usize>().unwrap();
+
+        // allocate matrix for the adjacency matrix
+        let mut adj = vec![vec![0; cols]; rows];
+
+        // and fill it with the edges from the mtx file
+        for line in lines {
+            let mut iter = line.split_whitespace();
+            let i = iter.next().unwrap().parse::<usize>().unwrap();
+            let j = iter.next().unwrap().parse::<usize>().unwrap();
+
+            // for undirected edges in CPDAGs:
+            // let edge_type = iter.next().unwrap().parse::<i8>().unwrap();
+
+            adj[i - 1][j - 1] = 1;
+        }
+
+        PDAG::from_vecvec(adj)
+    }
+
+    /// Takes two names, like `g_true="DAG1"` and `g_guess="DAG2"` and returns a Testcase, loading from the corresponding `../testgraphs/{g_true}.mtx` files
+    fn test(g_true_name: &str, g_guess_name: &str) -> Testcase {
+        // get the root of the project
+        let root = std::env::var("CARGO_MANIFEST_DIR").unwrap();
+        // get the parent directory of the project
+        let root_parent = std::path::Path::new(&root).parent().unwrap();
+        // get the child dir "testgraphs"
+        let testgraphs = root_parent.join("testgraphs");
+
+        // load the true and guess graphs
+        let g_true = load_pdag_from_mtx(
+            &testgraphs
+                .join(format!("{}.mtx", g_true_name))
+                .to_str()
+                .unwrap(),
+        );
+        let g_guess = load_pdag_from_mtx(
+            &testgraphs
+                .join(format!("{}.mtx", g_guess_name))
+                .to_str()
+                .unwrap(),
+        );
+
+        // get deterministic seed by hashing the two graph names
+        let mut hasher = std::collections::hash_map::DefaultHasher::new();
+        g_true_name.hash(&mut hasher);
+        g_guess_name.hash(&mut hasher);
+        let seed = hasher.finish();
+
+        // using rand_chacha to sample nodes with seed because it is reproducible across platforms
+        let mut rng = rand_chacha::ChaCha8Rng::seed_from_u64(seed);
+
+        let t = rng.gen_range(0..g_true.n_nodes as usize);
+        let mut y = rng.gen_range(0..g_true.n_nodes as usize);
+        if y == t {
+            y = (y + 1) % g_true.n_nodes as usize;
+        }
+        let z = g_guess.parents_of(t);
```

Comment:
This is a very special z.
Can we randomly choose between one of the four:
a) parents, b) ancestors, c) non-descendants, d) a random subset of nodes that are not t and y and of size uniformly chosen between 2 and the size of the ancestor set

Reply:
I don't think we have a good technique to get non-descendants (= possible ancestors?) for now, would have to make a new file + function for that.